txstratum/manager.py

```diff
@@ -264,7 +279,10 @@ def _job_timeout_if_possible(self, job: TxJob) -> None:
         self.log.info('TxJob timeout', job=job.to_dict())
         self.txs_timeout += 1
         job.status = JobStatus.TIMEOUT
-        self.tx_queue.remove(job)
+        try:
+            self.tx_queue.remove(job)
+        except ValueError:
```

Comment:
If the reason for this exception really is that a solution was found right before executing the timeout (and not that maybe we had 2 timeouts scheduled for the same job), then the code below shouldn't be executed, right? In the `submit_solution` method we already execute `stop_mining_tx` and `schedule_job_clean_up`. It doesn't seem to harm because both methods are ok to be executed twice but may not be needed.

Anyway, I think we should handle this earlier in the method, otherwise we are changing the job status to the wrong status and wrongly adding 1 to the timeout count.

Reply:
> If the reason for this exception really is that a solution was found right before executing the timeout (and not that maybe we had 2 timeouts scheduled for the same job)

Yeah, I'm not really sure what is the cause for the exception here.

I don't think it's because of a solution found, neither because of 2 timeouts, because both conditions would be caught in the first if:
```python
        if job.status in JobStatus.get_after_mining_states():
            return
```

I think the real cause for this exception was the other exception that we are also fixing in this PR. By looking at the code where the other exception happens, you will see that it prevented all those things from running in the `submit_solution` method:

```python
            self.schedule_job_clean_up(tx_job)
            # Mark as solved, stop mining it, and propagate if requested.
            tx_job.mark_as_solved(nonce=nonce, timestamp=job.get_object().timestamp)
            self.stop_mining_tx(tx_job)
```

So, we not only would not be marked the job as `DONE`, but also we wouldn't schedule its cleaning to remove the timeout scheduled, and we would not stop other miners from mining it, which is the worse thing.

So, after this other exception happened, the timeout would still be triggered even for a tx that was solved, because it was not marked as solved. But it could have been removed from the queue. All these things would generate the exception in this timeout code.

>  then the code below shouldn't be executed, right? In the submit_solution method we already execute stop_mining_tx and schedule_job_clean_up. It doesn't seem to harm because both methods are ok to be executed twice but may not be needed.

Because of the explanation above, if things work ok, they won't be executed twice. But if something unexpected happens, we would make sure they run at least during the timeout.

> Anyway, I think we should handle this earlier in the method, otherwise we are changing the job status to the wrong status and wrongly adding 1 to the timeout count.

I think the first if in the method is protecting agaisnt this. An exception would be raised here only in strage situations. I'm thinking even in logging it as an error, or not treating this at all (just the other one).

But I guess handling this earlier doesn't hurt.

Reply:
I'll change the log to an error

Reply:
I agree with the explanation and the possible root cause for the problem. Maybe you should change the PR description, it makes more sense with what you commented here.

Reply:
I agree with the explanation and the possible root cause for the problem. Maybe you should change the PR description, it makes more sense with what you commented here.